### ISSUE TITLE
Wrong error message shown when a `_` is used in place of generic constant

### ISSUE THREAD

**Issue by InnocentusLime** [rust-lang/rust#84298]
Given the following code: 

```rust
static TEST : [i32; _] = [1i32, 2i32];
fn main() {}
```

The current output is:

```
error[E0658]: destructuring assignments are unstable
 --> src/main.rs:1:21
  |
1 | static TEST : [i32; _] = [1i32, 2i32];
  |                     ^
  |
  = note: see issue #71126 <https://github.com/rust-lang/rust/issues/71126> for more information
```

Ideally the output should look like:

```
error[E0747]: type provided when a constant was expected
 --> src/main.rs:3:16
  |
1 | static TEST : [i32; _] = [1i32, 2i32];
  |                     ^
  |
  = help: const arguments cannot yet be inferred with `_`
```
The sample code doesn't seem to be related to the `destructuring assignment` feature nonetheless.

This, although quite artificial looking, use case too causes the wrong looking output

```rust
fn test() -> [i32; _] { [2, 1] }
fn main() {}
```

Output:

```
error[E0658]: destructuring assignments are unstable
 --> src/main.rs:3:20
  |
3 | fn test() -> [i32; _] { [2, 1] }
  |                    ^
  |
  = note: see issue #71126 <https://github.com/rust-lang/rust/issues/71126> for more information
  = help: add `#![feature(destructuring_assignment)]` to the crate attributes to enable

error: in expressions, `_` can only be used on the left-hand side of an assignment
 --> src/main.rs:3:20
  |
3 | fn test() -> [i32; _] { [2, 1] }
  |                    ^ `_` not allowed here
```

Some code seems to print correct the error messages (kinda?):

```rust
struct T<const U : usize>([i32; U]);
fn test() -> T<_> { T([2, 1]) }
fn main() {}
```

Output

```
error[E0747]: type provided when a constant was expected
 --> src/main.rs:3:16
  |
3 | fn test() -> T<_> { T([2, 1]) }
  |                ^
  |
  = help: const arguments cannot yet be inferred with `_`

error[E0121]: the type placeholder `_` is not allowed within types on item signatures
 --> src/main.rs:3:16
  |
3 | fn test() -> T<_> { T([2, 1]) }
  |                ^ not allowed in type signatures
```

# Meta

This happens on stable, beta and nightly.

**Comment by tdelabro** [comment]
I would be interested in fixing this issues, but it will be my first contribution to the compiler so I may need some help.
I started looking at the code an this is what i understood:

The error you would like to see is emitted here, and I think it is after (or during) the conversion from ast to mir
https://github.com/rust-lang/rust/blob/5258a74c887f8ae14717e1f98b652b470877ce4e/compiler/rustc_typeck/src/astconv/generics.rs#L41-L45

The ones you get are emitted here:
https://github.com/rust-lang/rust/blob/e9cdcccfa8321363999a43ee30a18b3aadbc20de/compiler/rustc_ast_lowering/src/expr.rs#L1007-L1016
And here:
https://github.com/rust-lang/rust/blob/e9cdcccfa8321363999a43ee30a18b3aadbc20de/compiler/rustc_ast_lowering/src/expr.rs#L194-L203

This take place during the lowering of ast expressions, which I assume is executed before the first one and therefore prevent it's execution.

I see two possible solutions:
 - catch this misuse of the `_` symbol earlier in the process, possibly in https://github.com/rust-lang/rust/blob/master/compiler/rustc_ast_passes/src/ast_validation.rs. I guess it will block the ast lowering and came out as the only error.
 - fix what I think is a bug: `static TEST : [i32; _] = [1i32, 2i32];` is not doing an array destructuring but a simple assignment. I guess `[i32; _] ` is interpreted as the lef thand side of a destructuring expression, when it is actually a type. If we get rid of this behaviour the compilation will probably continue until it raise the `E0747` we are expecting.
 
 As I am very new to this codebase I would greatly appreciate a more experienced diagnostic on this issue :)

**Comment by BoxyUwU** [comment]
the following [code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=126ff2f15a1e1bd8c25fcd85b852a2ca) is the same issue I believe:
```rust
fn main() {
  let foo: [_; _] = [];
}
```

**Comment by estebank** [comment]
Current output:
```
error[E0121]: the placeholder `_` is not allowed within types on item signatures for static variables
 --> src/main.rs:3:21
  |
3 | static TEST : [i32; _] = [1i32, 2i32];
  |                     ^ not allowed in type signatures
  |
help: replace this with a fully-specified type
  |
3 - static TEST : [i32; _] = [1i32, 2i32];
3 + static TEST : [i32; 2] = [1i32, 2i32];
  |
```
```
error[E0121]: the placeholder `_` is not allowed within types on item signatures for return types
 --> src/main.rs:1:20
  |
1 | fn test() -> [i32; _] { [2, 1] }
  |              ------^-
  |              |     |
  |              |     not allowed in type signatures
  |              help: replace with the correct return type: `[i32; 2]`
```
```
error[E0121]: the placeholder `_` is not allowed within types on item signatures for return types
 --> src/main.rs:2:16
  |
2 | fn test() -> T<_> { T([2, 1]) }
  |              --^-
  |              | |
  |              | not allowed in type signatures
  |              help: replace with the correct return type: `T<2>`

```
```
error[E0282]: type annotations needed
 --> src/main.rs:2:23
  |
2 |     let foo: [_; _] = [];
  |                       ^^ cannot infer type
```
